### ISSUE TITLE
Cannot add worktree with Git version 2.16 due to post-checkout hook locking up

### ISSUE THREAD

**Issue by xoob** [git-lfs/git-lfs#2848]
`git-lfs-post-checkout(1)` hangs when creating a worktree from a bare repo with Git version >= 2.16. This does not happen with Git version 2.15.x. We have not tested whether or not this happens with non-bare repos.

Issue
------

The latest release, Git started executing hooks after creating the worktree folder, which it didn't in earlier versions.

> "git worktree add" learned to run the post-checkout hook, just like "git checkout" does, after the initial checkout. – https://github.com/git/git/commit/e87f9fc9d4ead0ec7f98c9493a9f2ad8c41cd6ee 

This actually addresses an issue where worktrees would only contain the pointer files and needed a separate `git lfs checkout`. However, the hook call seems to be incompatible with Git LFS version 2.3.4, which locks up during the execution of `git-lfs-post-checkout(1)`:

    $ git worktree add -f -b temp1517404720 /tmp/temp1517404720 master &
    $ ps f -g $(ps -o sid= -p $!)
    PID  STAT   TIME COMMAND
    4003 SNs    0:00 /usr/bin/git worktree add -f -b temp1517404720 /tmp/temp1517404720 master
    4009 SN     0:00  \_ /bin/sh hooks/post-checkout 0000000000000000000000000000000000000000 609a31113bd96321699b354860d51d4b19f66926 1
    4010 SN     0:00      \_ git lfs post-checkout 0000000000000000000000000000000000000000 609a31113bd96321699b354860d51d4b19f66926 1
    4011 SNl    0:00          \_ /usr/bin/git-lfs post-checkout 0000000000000000000000000000000000000000 609a31113bd96321699b354860d51d4b19f66926 1
    $ kill $!

`git-lfs-post-checkout` hangs indefinitely.

Workaround
----------

To work around the issue, we found that disabling the checkout hooks with
`--no-checkout` prevents the lock up. This leaves the worktree directory
empty. To populate the files, we then execute `git-checkout(1)` inside the
worktree directory, at which point the hooks succeed.

    $ git worktree add -f -b temp1517404721 /tmp/temp1517404721 master
    $ cd /tmp/temp1517404721
    $ git checkout HEAD

System Details
------------
```
git-lfs/2.3.4 (GitHub; linux amd64; go 1.8.3)
git version 2.16.1

Endpoint=[redacted] (auth=basic)
LocalWorkingDir=
LocalGitDir=[redacted].git
LocalGitStorageDir=[redacted].git
LocalMediaDir=[redacted].git/lfs/objects
LocalReferenceDir=
TempDir=[redacted].git/lfs/tmp
ConcurrentTransfers=3
TusTransfers=false
BasicTransfersOnly=true
SkipDownloadErrors=false
FetchRecentAlways=false
FetchRecentRefsDays=7
FetchRecentCommitsDays=0
FetchRecentRefsIncludeRemotes=true
PruneOffsetDays=3
PruneVerifyRemoteAlways=false
PruneRemoteName=origin
LfsStorageDir=[redacted].git/lfs
AccessDownload=basic
AccessUpload=basic
DownloadTransfers=basic
UploadTransfers=basic
GIT_LFS_VERSION=2.3.4
GIT_VERSION=2.16
git config filter.lfs.process = "git-lfs filter-process"
git config filter.lfs.smudge = "git-lfs smudge -- %f"
git config filter.lfs.clean = "git-lfs clean -- %f"
```

**Comment by larsxschneider** [comment]
@xoob: Thanks for the report! I tried a number of different things but I wasn't able to reproduce the behaviour so far.

Can you run the following and post the (redacted) output?
```
export GIT_TRACE=1 
git worktree add -f -b temp1517404720 /tmp/temp1517404720 master
```

Further, can you check the used Git LFS version (maybe you have another version somewhere else in your path)?
```
/usr/bin/git-lfs version
```

Do you use the Git LFS locking feature?
Is there anything special about your Git/Git LFS setup or repo?

**Comment by xoob** [comment]
@larsxschneider Thanks for looking into it. Yes, we are running in a batch environment which is unusual. I will set up a test case to reproduce this outside of our system and report back. It’s going to take me a couple of days thought since I’m gone over the weekend.

**Comment by xoob** [comment]
@larsxschneider I've put together a test image to reproduce this problem. It installs Git and Git LFS, then creates an bare repo, and attempts to create a worktree. The 2.15 build works, the 2.16 build hangs.

## Test Script

Save this as `./Dockerfile`:

```dockerfile
FROM buildpack-deps:trusty

RUN set -ex \
    && apt-get update \
    && apt-get install -y --no-install-recommends \
        apt-transport-https \
        python-software-properties \
        software-properties-common \
    && apt-get clean

### Build Git

ARG GIT_VERSION

RUN set -ex \
    && apt-get purge -y --auto-remove git || : \
    && apt-get purge -y --auto-remove git-core || : \
    && buildDeps=' \
        libcurl4-openssl-dev \
        libexpat1-dev \
        libz-dev \
        libssl-dev \
    ' \
    && apt-get update \
    && apt-get install -y --no-install-recommends $buildDeps \
    && apt-get clean \
    && curl -fSL -o git.tar.gz "https://www.kernel.org/pub/software/scm/git/git-${GIT_VERSION}.tar.gz" \
    && mkdir -p /usr/src/git \
    && tar -xzf git.tar.gz -C /usr/src/git --strip-components=1 \
    && rm git.tar.gz \
    && cd /usr/src/git \
    && make configure \
    && ./configure --prefix=/usr/local --without-tcltk \
    && { \
        echo "NO_GETTEXT = 1;"; \
        echo "NO_SVN_TESTS = 1;"; \
    } >config.mak \
    && make -j"$(nproc)" \
    && make install \
    && rm -r /usr/src/git \
    && git --version

### Install git-lfs

ENV GIT_LFS_VERSION=2.3.4

RUN set -ex \
    && curl -sL https://packagecloud.io/github/git-lfs/gpgkey | apt-key add - \
    && echo "deb https://packagecloud.io/github/git-lfs/ubuntu/ trusty main" >>/etc/apt/sources.list \
    && apt-get update \
    && apt-get download git-lfs=${GIT_LFS_VERSION}* \
    && dpkg -i --ignore-depends=git git-lfs_*.deb \
    && apt-get clean \
    && git lfs version

### Run Test Script

CMD set -ex \
    && git config --global user.email 'you@example.com' \
    && git config --global user.name 'Your Name' \
    \
    && mkdir -p /tmp/repo.git \
    && cd /tmp/repo.git \
    && git init --bare . \
    && git lfs install --local \
    && echo "creating empty branch 'master'..." \
    && true | git mktree | xargs git commit-tree | xargs git branch master \
    \
    && git lfs env \
    \
    && echo "running git-worktree-add..." \
    && GIT_TRACE=1 git worktree add -f -b temp1 /tmp/temp1 master \
    \
    && echo "done."
```

## Build images with 2.15 and 2.16

```console
$ docker build --build-arg GIT_VERSION=2.15.1 -t v2_15_1 .
$ docker build --build-arg GIT_VERSION=2.16.1 -t v2_16_1 .
```

## Run test cases

With the old Git 2.15:

```console
$ docker run --rm -ti v2_15_1

+ git lfs env
git-lfs/2.3.4 (GitHub; linux amd64; go 1.8.3)
git version 2.15.1

LocalWorkingDir=
LocalGitDir=/tmp/repo.git
LocalGitStorageDir=/tmp/repo.git
LocalMediaDir=/tmp/repo.git/lfs/objects
LocalReferenceDir=
TempDir=/tmp/repo.git/lfs/tmp
ConcurrentTransfers=3
TusTransfers=false
BasicTransfersOnly=false
SkipDownloadErrors=false
FetchRecentAlways=false
FetchRecentRefsDays=7
FetchRecentCommitsDays=0
FetchRecentRefsIncludeRemotes=true
PruneOffsetDays=3
PruneVerifyRemoteAlways=false
PruneRemoteName=origin
LfsStorageDir=/tmp/repo.git/lfs
AccessDownload=none
AccessUpload=none
DownloadTransfers=basic
UploadTransfers=basic
GIT_LFS_VERSION=2.3.4
git config filter.lfs.process = "git-lfs filter-process"
git config filter.lfs.smudge = "git-lfs smudge -- %f"
git config filter.lfs.clean = "git-lfs clean -- %f"

+ echo running git-worktree-add...
running git-worktree-add...

+ GIT_TRACE=1 git worktree add -f -b temp1 /tmp/temp1 master
17:21:34.414012 git.c:344               trace: built-in: git 'worktree' 'add' '-f' '-b' 'temp1' '/tmp/temp1' 'master'
17:21:34.414098 run-command.c:626       trace: run_command: 'branch' 'temp1' 'master'
17:21:34.414596 git.c:344               trace: built-in: git 'branch' 'temp1' 'master'
Preparing /tmp/temp1 (identifier temp1)
17:21:34.415212 run-command.c:626       trace: run_command: 'symbolic-ref' 'HEAD' 'refs/heads/temp1'
17:21:34.415845 git.c:344               trace: built-in: git 'symbolic-ref' 'HEAD' 'refs/heads/temp1'
17:21:34.416131 run-command.c:626       trace: run_command: 'reset' '--hard'
17:21:34.416738 git.c:344               trace: built-in: git 'reset' '--hard'
HEAD is now at af83a7b

+ echo done.
done.
```

And with the new Git 2.16:

```console
$ docker run --rm -ti v2_16_1

+ git lfs env
git-lfs/2.3.4 (GitHub; linux amd64; go 1.8.3)
git version 2.16.1

LocalWorkingDir=
LocalGitDir=/tmp/repo.git
LocalGitStorageDir=/tmp/repo.git
LocalMediaDir=/tmp/repo.git/lfs/objects
LocalReferenceDir=
TempDir=/tmp/repo.git/lfs/tmp
ConcurrentTransfers=3
TusTransfers=false
BasicTransfersOnly=false
SkipDownloadErrors=false
FetchRecentAlways=false
FetchRecentRefsDays=7
FetchRecentCommitsDays=0
FetchRecentRefsIncludeRemotes=true
PruneOffsetDays=3
PruneVerifyRemoteAlways=false
PruneRemoteName=origin
LfsStorageDir=/tmp/repo.git/lfs
AccessDownload=none
AccessUpload=none
DownloadTransfers=basic
UploadTransfers=basic
GIT_LFS_VERSION=2.3.4
git config filter.lfs.process = "git-lfs filter-process"
git config filter.lfs.smudge = "git-lfs smudge -- %f"
git config filter.lfs.clean = "git-lfs clean -- %f"

+ echo running git-worktree-add...
running git-worktree-add...

+ GIT_TRACE=1 git worktree add -f -b temp1 /tmp/temp1 master
17:25:17.289578 git.c:344               trace: built-in: git 'worktree' 'add' '-f' '-b' 'temp1' '/tmp/temp1' 'master'
17:25:17.289653 run-command.c:627       trace: run_command: 'branch' 'temp1' 'master'
17:25:17.290154 git.c:344               trace: built-in: git 'branch' 'temp1' 'master'
Preparing /tmp/temp1 (identifier temp1)
17:25:17.290894 run-command.c:627       trace: run_command: 'symbolic-ref' 'HEAD' 'refs/heads/temp1'
17:25:17.291491 git.c:344               trace: built-in: git 'symbolic-ref' 'HEAD' 'refs/heads/temp1'
17:25:17.291769 run-command.c:627       trace: run_command: 'reset' '--hard'
17:25:17.292436 git.c:344               trace: built-in: git 'reset' '--hard'
HEAD is now at 17db062
17:25:17.293213 run-command.c:627       trace: run_command: 'hooks/post-checkout' '0000000000000000000000000000000000000000' '17db0622cb8bfb0f6da50e782b91910564d31577' '1'
17:25:17.293990 git.c:576               trace: exec: 'git-lfs' 'post-checkout' '0000000000000000000000000000000000000000' '17db0622cb8bfb0f6da50e782b91910564d31577' '1'
17:25:17.294009 run-command.c:627       trace: run_command: 'git-lfs' 'post-checkout' '0000000000000000000000000000000000000000' '17db0622cb8bfb0f6da50e782b91910564d31577' '1'
17:25:17.296000 trace git-lfs: run_command: 'git' version
17:25:17.297229 trace git-lfs: run_command: 'git' config -l
17:25:17.298084 trace git-lfs: Error finding .gitattributes: stat : no such file or directory
^C
```

**Comment by larsxschneider** [comment]
Nice. Looking into it...

**Comment by larsxschneider** [comment]
Hmm... you create a branch with an empty commit tree there with lower level commands. The higher level Git commands don't even allow that I think. All this wasn't a problem with Git 2.15 as the `post-checkout` hook was not executed for worktrees. Git 2.16 changes that.  In the `post-checkout` hook we [walk the Git repo](https://github.com/git-lfs/git-lfs/blob/0db4437efa9e2f0714e4b4f2061535f19fcbe99b/git/attribs.go#L149) and I assume we don't handle the case of a branch with a single commit that contains an empty tree...

**Comment by xoob** [comment]
The empty tree is just there for the test script, so Git has a first commit that it can create the worktree from. But I see what you mean. The same thing happened with real bare repos that contain normal files + LFS files + `.gitattributes` too. 

I can update the script so the bare repo has more realistic contents if that would help. Let me know if I should. 

Does the last message `Error finding .gitattributes` have anything to do with it or is it just informational?

**Comment by larsxschneider** [comment]
I looked again at it and what I wrote earlier was wrong. While we could improve the error handling in Git LFS a bit here, the real problem is in Git. The `post-checkout` hook is not aware of the worktree and consequently Git LFS is not aware of it either. That means Git LFS uses the main worktree for its `post-checkout` operation. That's wrong but it doesn't cause visible errors. In your case there is no main worktree as you have a bare repo. That makes the error visible. I'll post a patch to the Git core mailing list to fix the bug and then I will try to find a workaround for Git LFS.

You can try my Git core patch with this dockerfile:
```
FROM buildpack-deps:trusty

RUN set -ex \
    && apt-get update \
    && apt-get install -y --no-install-recommends \
        apt-transport-https \
        build-essential \
        gettext \
        libcurl4-openssl-dev \
        libexpat-dev \
        libssl-dev \
        libz-dev \
        python-software-properties \
        software-properties-common \
    && apt-get clean \
    && mkdir -p /usr/src/git \
    && git clone https://github.com/larsxschneider/git /usr/src/git

### Build Git

RUN set -ex \
    && cd /usr/src/git \
    && git fetch \
    && git checkout -f origin/fix-worktree-add \
    && git show \
    && apt-get purge -y --auto-remove git || : \
    && apt-get purge -y --auto-remove git-core || : \
    && make configure \
    && ./configure --prefix=/usr \
    && make -j"$(nproc)" \
    && make install

### Install git-lfs

ENV GIT_LFS_VERSION=2.3.4

RUN set -ex \
    && git version \
    && curl -sL https://packagecloud.io/github/git-lfs/gpgkey | apt-key add - \
    && echo "deb https://packagecloud.io/github/git-lfs/ubuntu/ trusty main" >>/etc/apt/sources.list \
    && apt-get update \
    && apt-get download git-lfs=${GIT_LFS_VERSION}* \
    && dpkg -i --ignore-depends=git git-lfs_*.deb \
    && apt-get clean \
    && git lfs version

### Run Test Script

CMD set -ex \
    && git config --global user.email 'you@example.com' \
    && git config --global user.name 'Your Name' \
    \
    && mkdir -p /tmp/repo.git \
    && cd /tmp/repo.git \
    && git init --bare . \
    && git lfs install --local \
    && echo "creating empty branch 'master'..." \
    && true | git mktree | xargs git commit-tree | xargs git branch master \
    \
    && git lfs env \
    \
    && echo "running git-worktree-add..." \
    && GIT_TRACE=1 git worktree add -f -b temp1 /tmp/temp1 master \
    \
    && echo "done."
```

**Comment by xoob** [comment]
Thanks so much for your help. Confirmed this works for me and prevents the command from getting stuck. I'm going to follow the [discussion on the mailing list](https://public-inbox.org/git/CAPig+cQ7MRQVT-1yNF8t6RMHNrgMtYvzOToamaLO==ymxTtLYg@mail.gmail.com/T/) for updates. 
Should I close this issue or keep it open until it's fixed in Git?

P.S. had to add this back to to the `Dockerfile` or else it wouldn't build:

```
    && make configure \
    && ./configure --prefix=/usr --without-tcltk \
    && { \
        echo "NO_GETTEXT = 1;"; \
    } >config.mak \
    && make -j"$(nproc)" \
```

**Comment by ttaylorr** [comment]
> Should I close this issue or keep it open until it's fixed in Git?

Nice work debugging this to the both of you! 👏 

I think that we can leave this issue open until the issue is resolved upstream, at which point I think it would make sense to close.

**Comment by larsxschneider** [comment]
Thanks 😄 

@ttaylorr I've added two new labels to track these kind of issues. If you don't like the labels or if you think I should not interfere that way with your issue management then please tell me. I won't be offended and I will delete them again!

**Comment by ttaylorr** [comment]
@larsxschneider Thanks! These labels are great.

**Comment by larsxschneider** [comment]
The fix made it into `next` and is scheduled for `master` (IOW next Git release): https://github.com/git/git/commit/a4bf1e3c2e4d0583a1822419fb54a566e6e46162

**Comment by larsxschneider** [comment]
📣 **The fix was released with Git 2.17** 🎉
If you experience this problem, then please update to Git 2.17 or higher.
